gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/StructuredQuery.java

```diff
@@ -714,32 +687,34 @@ B addGroupBy(String property, String... others) {
       return self();
     }
 
-    B mergeFrom(DatastoreV1.Query queryPb) {
+    B mergeFrom(com.google.datastore.v1beta3.Query queryPb) {
       if (queryPb.getKindCount() > 0) {
         kind(queryPb.getKind(0).getName());
       }
-      if (queryPb.hasStartCursor()) {
+      if (!queryPb.getStartCursor().isEmpty()) {
         startCursor(new Cursor(queryPb.getStartCursor()));
       }
-      if (queryPb.hasEndCursor()) {
+      if (!queryPb.getEndCursor().isEmpty()) {
         endCursor(new Cursor(queryPb.getEndCursor()));
       }
-      if (queryPb.hasOffset()) {
-        offset(queryPb.getOffset());
-      }
+      offset(queryPb.getOffset());
       if (queryPb.hasLimit()) {
-        limit(queryPb.getLimit());
+        limit(queryPb.getLimit().getValue());
       }
       if (queryPb.hasFilter()) {
-        filter(Filter.fromPb(queryPb.getFilter()));
+        Filter currFilter = Filter.fromPb(queryPb.getFilter());
+        if (currFilter != null) {
+          filter(currFilter);
+        }
       }
-      for (DatastoreV1.PropertyOrder orderByPb : queryPb.getOrderList()) {
+      for (com.google.datastore.v1beta3.PropertyOrder orderByPb : queryPb.getOrderList()) {
         addOrderBy(OrderBy.fromPb(orderByPb));
       }
-      for (DatastoreV1.PropertyExpression projectionPb : queryPb.getProjectionList()) {
+      for (com.google.datastore.v1beta3.Projection projectionPb
+           : queryPb.getProjectionList()) {
         addProjection(Projection.fromPb(projectionPb));
       }
-      for (DatastoreV1.PropertyReference groupByPb : queryPb.getGroupByList()) {
+      for (com.google.datastore.v1beta3.PropertyReference groupByPb : queryPb.getDistinctOnList()) {
```

Comment:
No, this is different. If a user wants a standard distinct selection, they have to specify all of them (i.e. `SELECT DISTINCT a, b FROM K`).

However, if you wanted the distinct `a` values with the "first" value of `b`, you could write:
`SELECT DISTINCT ON(a) a, b FROM K ORDER BY a, b

The reason this is different than `GROUP BY` is that you are allowed to put values in the `SELECT` clause that aren't in the `DISTINCT ON` clause and the orders are applied *before* the `DISTINCT ON` whereas orders are applied *after* `GROUP BY`.

We may eventually add aggregation functions (and with it `GROUP BY`), but they are distinct from `DISTINCT ON`.

See the [postgres docs](http://www.postgresql.org/docs/9.0/static/sql-select.html#SQL-DISTINCT) for more info.